### PR TITLE
[libc++] Unify the benchmarks with the test suite

### PR DESCRIPTION
**Pull request by ldionne**
Instead of building the benchmarks separately via CMake and running them
separately from the test suite, this patch merges the benchmarks into
the test suite and handles both uniformly.

As a result:
- It is now possible to run individual benchmarks like we run tests
  (e.g. using libcxx-lit), which is a huge quality-of-life improvement.

- The benchmarks will be run under exactly the same configuration as
  the rest of the tests, which is a nice simplification. This does
  mean that one has to be careful to enable the desired optimization
  flags when running benchmarks, but that is easy with e.g.
  `libcxx-lit <...> --param optimization=speed`.

- Benchmarks can use the same annotations as the rest of the test
  suite, such as `// UNSUPPORTED` & friends.

When running the tests via `check-cxx`, we only compile the benchmarks
because running them would be too time consuming. This introduces a bit
of complexity in the testing setup, and instead it would be better to
allow passing a --dry-run flag to GoogleBenchmark executables, which is
the topic of https://github.com/google/benchmark/issues/1827.

I am not really satisfied with the layering violation of adding the
%{benchmark_flags} substitution to cmake-bridge, however I believe
this can be improved in the future.

### CODE REVIEW

**Comment by ldionne** [comment]
Review note: I will split the first commit into a separate PR once I have addressed all the issues that arise from running the benchmarks as part of the test suite.

File: libcxx/docs/TestingLibcxx.rst [hunk 1]
```diff
@@ -438,48 +442,20 @@ Libc++ contains benchmark tests separately from the test of the test suite.
...
-
-  $ ninja -C build cxx-benchmarks
-
-This will build all of the benchmarks under ``<libcxx>/test/benchmarks`` to be
-built against the just-built libc++. The compiled tests are output into
-``build/libcxx/test/benchmarks``.
+The benchmarks are located under ``libcxx/test/benchmarks``. Running a benchmark
+works in the same way as running a test. Both the benchmarks and the tests share
+the same configuration, so make sure to enable the relevant optimization level
+when running the benchmarks.
```
Comment (mordante) [inline libcxx/docs/TestingLibcxx.rst:451]: This feels really error prone to me and might run benchmarks at lower optimization levels than expected.

**Review by mordante — COMMENTED** [review]
After this patch were are the benchmark executables stored? When benchmarking I typically want to run the before and after version.

Are the benchmarks still executed in the benchmark CI configuration?

**Comment by ldionne** [comment]
> After this patch were are the benchmark executables stored? When benchmarking I typically want to run the before and after version.

The benchmarks are stored in the same location as a normal test executable is stored. For example: `<build>/test/benchmarks/algorithms/Output/for_each.bench.cpp.dir/t.tmp.exe`.

However, the biggest benefit of this patch is that you don't need to care anymore. You can simply run a benchmark like you run a test, with `libcxx-lit <build> path/to/benchmark.bench.cpp`. So you don't need to know where exactly the executable is, just like when you run tests.

> 
> Are the benchmarks still executed in the benchmark CI configuration?

I removed that CI configuration. After this patch, the benchmarks are always dry-run as part of the test suite, but we don't have a dedicated CI job that runs the full benchmarks. We were not gaining anything from having that since we never looked at those results. Instead, as a follow-up I would like to work on setting up a dedicated machine that we can run the benchmarks on and ideally even have something to prevent regressions. That CI job could be triggered by a comment in the PR or something like that: we probably don't want to run it all the time since we'd just be wasting cycles.

File: libcxx/docs/TestingLibcxx.rst (hunk 1 above)
Comment (ldionne) [reply libcxx/docs/TestingLibcxx.rst:451]: I feel you, but it still seems like the right thing to do. We want to be able to run the benchmarks under arbitrary optimization levels so we can e.g. evaluate the performance of an algorithm when optimizations are disabled. This calls for not hardcoding any optimization level when running the benchmarks, just like we don't when we run the tests.

**Review by philnik777 — COMMENTED** [review]
I really like this. Not having to figure out how to build and run the benchmarks every time would be quite nice. It's also quite nice to be able to run the benchmarks in different configurations more easily.

**Comment by ldionne** [comment]
> I do care. I sometimes like to compare benchmarks from different iterations. Currently when I build a benchmark I get a message where the executable is stored. The new approach makes harder to find the location of the executable. I wonder whether we can show a message where the benchmark is stored, maybe always print the location in `format.py`?

If you run the test suite verbosely (I like to use Lit's `--show-all` option), it shows where the executables are created. That way you can copy the executable to the location of your choosing. What's nice is that we now handle benchmarks consistently with all other tests, so this process also applies to any test executable you might want to copy around.

**Review by philnik777 — APPROVED** [review]

**Comment by winner245** [comment]
It is a great new feature which allows us to run benchmark tests in the same way we run a normal test. This indeed simplify the benchmark testing process. I wish I had seen this patch earlier, as the old method of building cxx-benchmarks seemed to have just stopped working, which left me confused until I discovered this patch.

**Pull request merged**